oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2TokenIntrospectionAuthenticationProvider.java

```diff
@@ -0,0 +1,178 @@
+import org.springframework.security.oauth2.server.authorization.OAuth2Authorization;
+import org.springframework.security.oauth2.server.authorization.OAuth2AuthorizationService;
+import org.springframework.security.oauth2.server.authorization.TokenType;
+import org.springframework.security.oauth2.server.authorization.Version;
+import org.springframework.security.oauth2.server.authorization.client.RegisteredClient;
+import org.springframework.security.oauth2.server.authorization.token.TokenExpirationValidator;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * An {@link AuthenticationProvider} implementation for OAuth 2.0 Token Introspection.
+ *
+ * @author Gerardo Roza
+ * @since 0.0.4
+ * @see OAuth2TokenIntrospectionAuthenticationToken
+ * @see OAuth2AuthorizationService
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7662#section-2.1">Section 2.1 - Introspection Request</a>
+ */
+public class OAuth2TokenIntrospectionAuthenticationProvider implements AuthenticationProvider {
+	private final OAuth2AuthorizationService authorizationService;
+	private final Collection<JwtDecoder> jwtDecoders;
+
+	/**
+	 * Constructs an {@code OAuth2TokenIntrospectionAuthenticationProvider} using the provided parameters.
+	 *
+	 * @param authorizationService the authorization service
+	 * @param jwtDecoders all available decoders that might be used to parse the token as a JWT-encoded token
+	 */
+	public OAuth2TokenIntrospectionAuthenticationProvider(OAuth2AuthorizationService authorizationService,
+			Collection<JwtDecoder> jwtDecoders) {
+		Assert.notNull(authorizationService, "authorizationService cannot be null");
+		this.authorizationService = authorizationService;
+		this.jwtDecoders = jwtDecoders;
+	}
+
+	@Override
+	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+		OAuth2TokenIntrospectionAuthenticationToken tokenIntrospectionAuthentication = (OAuth2TokenIntrospectionAuthenticationToken) authentication;
+
+		OAuth2ClientAuthenticationToken clientPrincipal = getAuthenticatedClientElseThrowInvalidClient(
+				tokenIntrospectionAuthentication);
+		RegisteredClient registeredClient = clientPrincipal.getRegisteredClient();
+
+		TokenType tokenType = null;
+
+		try {
+			String tokenTypeHint = tokenIntrospectionAuthentication.getTokenTypeHint();
+			if (StringUtils.hasText(tokenTypeHint)) {
+				if (TokenType.REFRESH_TOKEN.getValue().equals(tokenTypeHint)) {
+					tokenType = TokenType.REFRESH_TOKEN;
+				} else if (TokenType.ACCESS_TOKEN.getValue().equals(tokenTypeHint)) {
+					tokenType = TokenType.ACCESS_TOKEN;
+				} else if (TokenType.ID_TOKEN.getValue().equals(tokenTypeHint)) {
+					tokenType = TokenType.ID_TOKEN;
+				} else {
+					throw new OAuth2AuthenticationException(new OAuth2Error(OAuth2ErrorCodes2.UNSUPPORTED_TOKEN_TYPE));
```

Comment:
I decided to leave this functionality as we have for the Token Revocation endpoint (as suggested), but just to be clear, I think this does not follow the specs correctly (more info in this issue I created: https://github.com/spring-projects-experimental/spring-authorization-server/issues/174) ,thus I think we'll have to update this logic :)